Feat/reflection serializer

piotrm50 opened the pull request:
# Description of change

Implementation of reflection based serialization, which uses MarshalUtil to handle writing and reading byte buffers.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Implemented unit tests and benchmarks inside the package.


## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

piotrm50 commented:
I also ran basic benchmarks to compare results of current solution implemented in Goshimmer and newly implemented reflection based approach. The serialized structures were different, that's why we can't compare results one-to-one, however this can give an overall idea of the difference:

Reflection based approach, serializing sample struct used in unit tests:
```
cpu: Intel(R) Core(TM) i9-9940X CPU @ 3.30GHz
BenchmarkSerialzer
BenchmarkSerialzer-28    	  230144	     10642 ns/op
```

Current GoShimmer approach, serializing sample transaction message:
```
cpu: Intel(R) Core(TM) i9-9940X CPU @ 3.30GHz
BenchmarkMessageToBytesLegacy
BenchmarkMessageToBytesLegacy-28    	42828028	        25.78 ns/op
PASS
```

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+			elemBytes, err := buffer.ReadBytes(bytesRead)
+			readOffset = buffer.ReadOffset()
+
+			elemBytesString := typeutils.BytesToString(elemBytes)
+			if err != nil {
+				return nil, err
+			}
+
+			if _, seenAlready := elementsMap[elemBytesString]; fieldMetadata.NoDuplicates && seenAlready {
+				continue
+			}
+
```
Check first whether the duplicate check is enforced and then only do this expensive byte-to-string thing.

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+			}{elem, elemBytes})
+			elementsMap[elemBytesString] = customtypes.Void
+		}
+		if fieldMetadata.LexicalOrder {
```
Check the lexical order up on each iteration instead of after having read already all elements, as you know during the iteration already, whether the ordering is violated by comparing to just the previous element.

piotrm50 replied:
The question is what behavior to expect when the order is violated - currently it deserializes whatever order is provided, and then sorts the elements. Should it return an error instead?

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		}
+		if fieldMetadata.LexicalOrder {
+			// sort inputs
+			sort.Slice(elements, func(i, j int) bool {
```
You can get rid of `sort.Slice` as you will `bytes.Compare()` directly within the read `for` loop

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+func (m *SerializationManager) deserializeArray(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	restoredArray := reflect.New(valueType).Elem()
+	arrayLen := valueType.Len()
+	if valueType.Elem().Kind() == reflect.Uint8 {
+		// special handling of byte array to optimize execution
+		bytesRead, err := buffer.ReadBytes(arrayLen)
+		if err != nil {
+			return nil, err
+		}
+		for i := 0; i < arrayLen; i++ {
+			restoredArray.Index(i).Set(reflect.ValueOf(bytesRead[i]))
+		}
+	} else {
+		for i := 0; i < arrayLen; i++ {
+			elem, err := m.DeserializeType(valueType.Elem(), fieldMetadata, buffer)
+			if err != nil {
+				return nil, err
+			}
+			restoredArray.Index(i).Set(reflect.ValueOf(elem))
+		}
+	}
+	return restoredArray.Interface(), nil
```
Remove else clause and do an early return in the first condition.

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+}
+
+func (m *SerializationManager) tryBuiltInDeserializer(p reflect.Value, structType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (processed bool, err error) {
+	if structType.Implements(binaryUnarshallerType) {
+		// if the struct implements encoding.BinaryUnmarshaller
+		// read number of bytes the struct was serialized to
+		var structSize int
+		var structBytes []byte
+		structSize, err = ReadLen(fieldMetadata.LengthPrefixType, buffer)
+		if err != nil {
+			return true, err
+		}
+		structBytes, err = buffer.ReadBytes(structSize)
+		if err != nil {
+			return true, err
+		}
+		// unmarshal the structure using read bytes
+		restoredStruct := p.Interface().(encoding.BinaryUnmarshaler)
+		err = restoredStruct.UnmarshalBinary(structBytes)
+		if err != nil {
+			return true, err
+		}
+		return true, nil
+	} else if structType.Implements(autoserializerBinararyDeserializerType) {
+		// if the structure implements BinaryDeserializer
+		// number of bytes is not needed, because if the structure has been serialized correctly,
+		// it will deserialize correctly and use only as many bytes from the buffer as needed, leaving others untouched
+		restoredStruct := p.Interface().(BinaryDeserializer)
+		err = restoredStruct.DeserializeBytes(buffer, m, fieldMetadata)
+		if err != nil {
+			return true, err
+		}
+		return true, nil
+	}
+	return
```
extract conditions into functions

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		var structBytes []byte
+		structSize, err = ReadLen(fieldMetadata.LengthPrefixType, buffer)
+		if err != nil {
+			return true, err
```
add `fmt.Errorf()` and don't return false

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		}
+		structBytes, err = buffer.ReadBytes(structSize)
+		if err != nil {
+			return true, err
```
add `fmt.Errorf()` and don't return false

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		restoredStruct := p.Interface().(encoding.BinaryUnmarshaler)
+		err = restoredStruct.UnmarshalBinary(structBytes)
+		if err != nil {
+			return true, err
```
add `fmt.Errorf()` and don't return false

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		// number of bytes is not needed, because if the structure has been serialized correctly,
+		// it will deserialize correctly and use only as many bytes from the buffer as needed, leaving others untouched
+		restoredStruct := p.Interface().(BinaryDeserializer)
+		err = restoredStruct.DeserializeBytes(buffer, m, fieldMetadata)
+		if err != nil {
+			return true, err
```
add `fmt.Errorf()` and don't return false

luca-moser replied:
Compress into `if err := ...; err != nil {`

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		buffer.WriteUint64(value.Uint())
+	case reflect.Float32:
+		if math.IsNaN(value.Float()) {
+			return fmt.Errorf("NaN float value")
```
create ``Err...` var

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		err = m.serializeArray(value, fieldMetadata, buffer)
+	case reflect.Slice:
+		err = m.serializeSlice(value, fieldMetadata, buffer)
+	case reflect.Map:
```
create `Err...` var

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+
+func (m *SerializationManager) serializeInterface(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	// write first byte only if AllowNil set to true
+	if value.IsNil() && fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(0))
+		return nil
+	} else if fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(1))
+	}
```
What happens if we don't allow `nil` but `value.IsNil()`?

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		return err
+	}
+	buffer.WriteUint32(encodedType)
+	err = m.SerializeValue(value.Elem(), fieldMetadata, buffer)
+	return err
+}
```
compress to `return ...`

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+func (m *SerializationManager) serializeStructure(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	valueType := value.Type()
+	// individual serialization for time.Time
+	if valueType == reflect.TypeOf(time.Time{}) {
+		buffer.WriteTime(value.Interface().(time.Time))
+		return nil
+	} else if processed, err := m.tryBuiltInSerializer(value, valueType, fieldMetadata, buffer); err != nil || processed {
+		// serialize using built-in serializer if it's available
+		return err
+	} else {
+		// serialize as regular struct
+		err = m.serializeFields(value, buffer)
+		return err
+	}
```
Use guard programming, i.e don't use if/else/else if, just early return (which you're doing anyway)

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+	}
+	for _, fieldMeta := range serializedFields {
+		fieldValue := value.Field(fieldMeta.idx)
+		err = m.SerializeValue(fieldValue, fieldMeta, buffer)
+		if err != nil {
+			break
```
compress

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+
+func (m *SerializationManager) serializePointer(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	// write first byte only if AllowNil set to true
+	if value.IsNil() && fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(0))
+		return nil
+	} else if fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(1))
+	}
```
Same code as in `serializeInterface()`

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+	}
+
+	// or else serialize as regular pointer
+	err = m.SerializeValue(value.Elem(), fieldMetadata, buffer)
+	return err
```
compress

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+	return err
+}
+
+func (m *SerializationManager) serializeSlice(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
```
see comments from the deserialization function for slices for comments

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		lengthUint32, err := buffer.ReadUint32()
+		return int(lengthUint32), err
+	default:
+		return 0, fmt.Errorf("unknown length prefix type %d", lenPrefixType)
```
create `Err...` var

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+// ValidateLength is used to make sure that the length of a collection is within bounds specified in struct tags.
+func ValidateLength(length int, minSliceLen int, maxSliceLen int) (err error) {
+	if length < minSliceLen {
+		err = fmt.Errorf("collection is required to have at least %d elements instead of %d", minSliceLen, length)
```
just return from here, create `Err...` var

luca-moser commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+		err = fmt.Errorf("collection is required to have at least %d elements instead of %d", minSliceLen, length)
+	}
+	if maxSliceLen > 0 && length > maxSliceLen {
+		err = fmt.Errorf("collection is required to have at most %d elements instead of %d", maxSliceLen, length)
```
same here

luca-moser commented on reflectionserializer/README.md:
```diff
@@ -0,0 +1,248 @@
+    * serialized bytes are serialized as a regular byte slice;
+    * during deserialization byte slice is read and passed to unmarshaler;
+* for custom collection types (e.g. `orderedmap.OrderedMap`), a special `BinarySerializer` and `BinaryDeserializer` is
+  used. Custom collection type needs to implement those interfaces internally;
```
```suggestion
* for custom collection types (e.g. `orderedmap.OrderedMap`), a special `BinarySerializer` and `BinaryDeserializer` is
  used. Custom collection types need to implement those interfaces internally;
```

luca-moser commented on reflectionserializer/README.md:
```diff
@@ -0,0 +1,248 @@
+| minLen          | int  | ensure that serialized collection has the specified min length.                                    | `slice` `orderedmap.OrderedMap`          |
+| maxLen          | int  | ensure that serialized collection has the specified max length                                     | `slice` `orderedmap.OrderedMap`          |
+| allowNil        | bool | whether given struct field can have nil value                                                      | `interface` `ptr`                        |
+| noDuplicates    | bool | whether a slice should have duplicate values removed during serialization and deserialization      | `slice`                                  |
+| lexicalOrder    | bool | whether a slice values should be lexicographically sorted before serialization and deserialization | `slice`                                  |
+| lenPrefixBytes  | int  | number of bytes to use for serialization of length of types with dynamic size.                     | `slice` `orderedmap.OrderedMap` `string` |
```
This should instead enforce the rules instead of auto. doing it; add specific tags for auto. removal/ordering.

piotrm50 replied:
Added two tags - `sort` and `skipDuplicates`. Previous tags return an error if order/duplicates requirement is violated.

luca-moser commented on reflectionserializer/README.md:
```diff
@@ -0,0 +1,248 @@
+If struct or pointer type implements `encoding.BinaryMarshaller` and `encoding.BinaryUnmarshaller`, then the method is
+used for serialization and deserialization. The interfaces are defined in standard Go libraries. The resulting bytes are
+serialized as a regular slice of bytes, therefore the serialized value is prefixed with the number of bytes the
+structure has been serialized to. During theserialization the length is used to read necessary number of bytes, and then
```
```suggestion
structure has been serialized to. During the serialization the length is used to read necessary number of bytes, and then
```

luca-moser commented on reflectionserializer/struct_field_cache.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// FieldMetadata contains information necessary to serialize and deserialize a struct field
+type FieldMetadata struct {
```
Perhaps makes sense to also add the field name so it can be used in the context where errors needs to be returned?

luca-moser commented on reflectionserializer/struct_field_cache.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// GetFields returns struct fields that are available for serialization. It caches the fields so consecutive calls for the same time can use previously extracted values.
+func (c *fieldCache) GetFields(structType reflect.Type) ([]FieldMetadata, error) {
```
Rename to more go-idiomatic `Fields()`

luca-moser commented on reflectionserializer/type_registry.go:
```diff
@@ -0,0 +1,62 @@
+	defer r.registryLock.RUnlock()
+	typeID, exists := r.typeToTypeID[t]
+	if !exists {
+		return 0, fmt.Errorf("type %v is not registered", t)
```
create `Err...` var

piotrm50 commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+}
+
+// SerializationManager stores TypeRegistry and is used to automatically serialize and deserialize structures.
+type SerializationManager struct {
```
Renamed to `Serializer`

piotrm50 commented on reflectionserializer/manager.go:
```diff
@@ -0,0 +1,795 @@
+
+func (m *SerializationManager) deserializeStructure(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	// handle time struct individually
+	if valueType == reflect.TypeOf(time.Time{}) {
+		restoredTime, err := buffer.ReadTime()
+		if err != nil {
+			return nil, err
+		}
+		return restoredTime, nil
+	}
+
```
Not really, because there we check `valueType.Kind()` of an object and here, knowing that it's a struct, we check if it's an instance of `time.Time` structure.

piotrm50 commented:
Addressed the suggestions and improved error messages overall. 

Added two tags - sort and skipDuplicates. Previous tags return an error if order/duplicates requirement is violated.

piotrm50 commented:
This one is closed as we went with another implementation of this feature.
